src/include/cart/api.h

```diff
@@ -685,98 +685,154 @@ crt_ep_abort(crt_endpoint_t *ep);
+	do {								\
+	CRT_GEN_GET_TYPE(seq) **e_ptrp = &ptr->CRT_GEN_GET_NAME(seq).ca_arrays;\
+	CRT_GEN_GET_TYPE(seq) *e_ptr = ptr->CRT_GEN_GET_NAME(seq).ca_arrays; \
+	uint64_t count = ptr->CRT_GEN_GET_NAME(seq).ca_count;		\
+	uint64_t i;							\
+	if (proc_op == CRT_PROC_DECODE)					\
+		*e_ptrp = NULL;						\
+	/* process the count of array first */				\
+	rc = crt_proc_uint64_t(proc, proc_op, &count);			\
+	if (unlikely(rc))						\
+		goto CRT_GEN_FAIL_LABEL(seq);				\
+	ptr->CRT_GEN_GET_NAME(seq).ca_count = count;			\
+	if (unlikely(count == 0))					\
+		break; /* goto next field */				\
+	if (proc_op == CRT_PROC_DECODE) {				\
+		D_ALLOC_ARRAY(e_ptr, (int)count);			\
+		if (unlikely(e_ptr == NULL)) {				\
+			rc = -DER_NOMEM;				\
+			goto CRT_GEN_FAIL_LABEL(seq);			\
 		}							\
-		/* process the elements of array */			\
-		for (i = 0; i < count; i++) {				\
-			rc = CRT_GEN_GET_FUNC(seq)(proc, proc_op, &e_ptr[i]); \
-			if (unlikely(rc)) {				\
-				if (DECODING(proc_op))			\
-					D_FREE(e_ptr);			\
-				goto out;				\
-			}						\
+		*e_ptrp = e_ptr;					\
+	}								\
+	/* process the elements of array */				\
+	for (i = 0; i < count && e_ptr != NULL; i++) {			\
+		rc = CRT_GEN_GET_FUNC(seq)(proc, proc_op, &e_ptr[i]);	\
+		if (unlikely(rc)) {					\
+			if (proc_op == CRT_PROC_DECODE)			\
+				ptr->CRT_GEN_GET_NAME(seq).ca_count = i;\
+			goto CRT_GEN_FAIL_LABEL(seq);			\
 		}							\
-		if (FREEING(proc_op))					\
-			D_FREE(e_ptr);					\
 	}								\
-	next_field_##r:,						\
-	BOOST_PP_IF(BOOST_PP_EQUAL(CRT_RAW, CRT_GEN_GET_KIND(seq)),	\
-	rc = crt_proc_memcpy(proc, proc_op, &ptr->CRT_GEN_GET_NAME(seq), \
+	if (proc_op == CRT_PROC_FREE)					\
+		D_FREE(*e_ptrp);					\
+	} while (0);
+
+#define CRT_GEN_FAIL_ARRAY(ptr, seq)					\
+	CRT_GEN_FAIL_LABEL(seq):					\
+	if (proc_op == CRT_PROC_DECODE) {				\
+		CRT_GEN_GET_TYPE(seq) **e_ptrp = &ptr->CRT_GEN_GET_NAME(seq).ca_arrays;\
+		CRT_GEN_GET_TYPE(seq) *e_ptr = ptr->CRT_GEN_GET_NAME(seq).ca_arrays; \
+		uint64_t count = ptr->CRT_GEN_GET_NAME(seq).ca_count;	\
+		uint64_t i;						\
+		/* process the elements of array */			\
+		for (i = 0; i < count && e_ptr != NULL; i++)		\
+			(void)CRT_GEN_GET_FUNC(seq)(proc, CRT_PROC_FREE,\
+						    &e_ptr[i]);		\
+		D_FREE(*e_ptrp);					\
+		ptr->CRT_GEN_GET_NAME(seq).ca_count = 0;		\
+	}
+
+#define CRT_GEN_PROC_RAW(ptr, seq)					\
+	rc = crt_proc_memcpy(proc, proc_op, &ptr->CRT_GEN_GET_NAME(seq),\
 			     sizeof(CRT_GEN_GET_TYPE(seq)));		\
 	if (unlikely(rc))						\
-		goto out;,						\
-	rc = CRT_GEN_GET_FUNC(seq)(proc, proc_op, &ptr->CRT_GEN_GET_NAME(seq));\
+		goto CRT_GEN_FAIL_LABEL(seq);
+
+#define CRT_GEN_FAIL_RAW(ptr, seq)					\
+	CRT_GEN_FAIL_LABEL(seq):;
+
+#define CRT_GEN_PROC_ALL(ptr, seq)					\
+	rc = CRT_GEN_GET_FUNC(seq)(proc, proc_op,			\
+				   &ptr->CRT_GEN_GET_NAME(seq));	\
 	if (unlikely(rc))						\
-		goto out;						\
-	))
+		goto CRT_GEN_FAIL_LABEL(seq);
+
+#define CRT_GEN_FAIL_ALL(ptr, seq)					\
```

Comment:
(style)  Macros with multiple statements should be enclosed in a do - while loop
(style)  macros should not use a trailing semicolon